src/scripts/coingeckoTop500.js

```diff
@@ -0,0 +1,279 @@
+
+function getCoingeckoApiUrl(chain) {
+  return `https://pro-api.coingecko.com/api/v3/coins/${DISPLAY_CHAIN_NAMES[chain]}/contract`
+}
+
+function getTokenListUrl(chain) {
+  return `https://tokens.coingecko.com/${COINGECKO_CHAINS[chain]}/all.json`
+}
+
+function getEmptyList(chain) {
+  return {
+    name: `Coingecko top ${TOP_TOKENS_COUNT} on ${COINGECKO_CHAINS[chain]}`,
+    logoURI:
+      'https://static.coingecko.com/s/thumbnail-007177f3eca19695592f0b8b0eabbdae282b54154e1be912285c9034ea6cbaf2.png',
+    keywords: ['defi'],
+    version: {
+      major: 0,
+      minor: 0,
+      patch: 0,
+    },
+  }
+}
+
+async function getTokenList(chain) {
+  const tokenListUrl = getTokenListUrl(chain)
+  const response = await fetch(tokenListUrl)
+  const data = await response.json()
+  return data.tokens
+}
+
+async function getTokenVolume(token, chain) {
+  try {
+    const headers = COINGECKO_API_KEY ? { 'X-Cg-Pro-Api-Key': COINGECKO_API_KEY } : {}
+
+    const url = `${getCoingeckoApiUrl(chain)}/${token.address}?localization=false`
+    const response = await fetch(url, { headers })
+    const data = await response.json()
+
+    // console.log(`[getTokenVolume] url: ${url}, chain: ${chain}`)
+    // console.log(`[getTokenVolume] token: ${token.symbol}, chain: ${chain}`)
+    // console.log(`[getTokenVolume] ${token.symbol} $${data.market_data.total_volume[VS_CURRENCY]}, chain: ${chain}`)
+    return data.market_data.total_volume[VS_CURRENCY] || 0
+  } catch (error) {
+    console.error(`Error fetching volume for token ${token.name} on chain ${chain}:`, error)
+    return 0
+  }
+}
+
+async function getCoingeckoTokenIds() {
+  try {
+    const headers = COINGECKO_API_KEY ? { 'X-Cg-Pro-Api-Key': COINGECKO_API_KEY } : {}
+
+    const url = `https://pro-api.coingecko.com/api/v3/coins/list?include_platform=true&status=active`
+    const response = await fetch(url, { headers })
+    return await response.json()
+  } catch (error) {
+    console.error(`Error fetching Coingecko's coin list`, error)
+    return []
+  }
+}
+
+/**
+ * Get a map of contract addresses to coingecko coin id.
+ *
+ * Result is stored in a local file since this shouldn't change often.
+ *
+ * Takes in one parameter forceUpdate, to ignore local content - if any.
+ */
+async function getCoingeckoTokenIdsMap(forceUpdate) {
+  const localFilePath = 'build/coingeckoTokenIdsMap.json'
+  let tokenIdsMap = COINGECKO_CHAINS_NAMES.reduce((acc, name) => ({ ...acc, [name]: {} }), {})
+
+  // Check if local file exists and forceUpdate is false
+  if (!forceUpdate && fs.existsSync(localFilePath)) {
+    try {
+      // Read the local file
+      const data = fs.readFileSync(localFilePath)
+      tokenIdsMap = JSON.parse(data)
+    } catch (error) {
+      console.error(`Error reading local file: ${error}`)
+    }
+  } else {
+    // Fetch the token IDs from Coingecko API
+    try {
+      const tokenIds = await getCoingeckoTokenIds()
```

Comment:
the token ids we only need them once? (next executions, uses the cached file?) what if there's new tokens?

Reply:
Hmm good point.
Although, when this is running on CI, the file won't be cached across runs since it's stored in the `build/` path.
I don't want to commit it.
More like a local cache in between runs.
But I can hold it only in memory and not store it in a file.